lib/portage/util/_eventloop/asyncio_event_loop.py

```diff
@@ -75,12 +80,17 @@ def _close(self):
             self._closing = False
 
     async def _close_main(self):
+        tasks = []
+        tasks.append(asyncio.ensure_future(mod_custom._async_finalize()))
```

Comment:
It feels a bit like an abstraction break but then again, the whole point of custom is it could do anything, so it's probably fine.

Reply:
We can add general purpose async atexit hooks, which we can use for both mod_custom and socks5 proxy.

Reply:
I've added a general purpose `portage.process.run_coroutine_exitfuncs()` function to handle it:
```diff
diff --git a/lib/portage/elog/mod_custom.py b/lib/portage/elog/mod_custom.py
index 1a74acdc95..0392c7afa7 100644
--- a/lib/portage/elog/mod_custom.py
+++ b/lib/portage/elog/mod_custom.py
@@ -26,4 +26,5 @@ def _get_procs() -> list[tuple[portage.process.MultiprocessingProcess, asyncio.F
     if _proc_refs is None or _proc_refs.pid != portage.getpid():
         _proc_refs = types.SimpleNamespace(pid=portage.getpid(), procs=[])
+        portage.process.atexit_register(_async_finalize)
     return _proc_refs.procs
 
diff --git a/lib/portage/process.py b/lib/portage/process.py
index b652e32942..0fedfea4ab 100644
--- a/lib/portage/process.py
+++ b/lib/portage/process.py
@@ -227,4 +227,26 @@ def run_exitfuncs():
 
 
+async def run_coroutine_exitfuncs():
+    """
+    This is the same as run_exitfuncs but it uses asyncio.iscoroutinefunction
+    to check which functions to run. It is called by the AsyncioEventLoop
+    _close_main method just before the loop is closed.
+    """
+    tasks = []
+    for index, (func, targs, kargs) in reversed(list(enumerate(_exithandlers))):
+        if asyncio.iscoroutinefunction(func):
+            del _exithandlers[index]
+            tasks.append(asyncio.ensure_future(func(*targs, **kargs)))
+    exc_info = None
+    for task in tasks:
+        try:
+            await task
+        except Exception:
+            traceback.print_exc()
+            exc_info = sys.exc_info()
+    if exc_info is not None:
+        raise exc_info[1].with_traceback(exc_info[2])
+
+
 atexit.register(run_exitfuncs)
 
diff --git a/lib/portage/util/_eventloop/asyncio_event_loop.py b/lib/portage/util/_eventloop/asyncio_event_loop.py
index 50cc0ef22a..2b34b28028 100644
--- a/lib/portage/util/_eventloop/asyncio_event_loop.py
+++ b/lib/portage/util/_eventloop/asyncio_event_loop.py
@@ -82,5 +82,5 @@ class AsyncioEventLoop(_AbstractEventLoop):
     async def _close_main(self):
         tasks = []
-        tasks.append(asyncio.ensure_future(mod_custom._async_finalize()))
+        tasks.append(asyncio.ensure_future(portage.process.run_coroutine_exitfuncs()))
 
         # Even though this has an exit hook, invoke it here so that
```

Reply:
Now the socks5 stuff cleans up nicely in https://github.com/gentoo/portage/pull/1299.